cnd/src/swap_protocols/rfc003/ledger_state.rs

```diff
@@ -18,26 +25,136 @@ pub enum LedgerState<L: Ledger> {
         htlc_location: L::HtlcLocation,
         deploy_transaction: L::Transaction,
         fund_transaction: L::Transaction,
+        asset: A,
     },
     Redeemed {
         htlc_location: L::HtlcLocation,
         deploy_transaction: L::Transaction,
         fund_transaction: L::Transaction,
         redeem_transaction: L::Transaction,
+        asset: A,
+        secret: Secret,
     },
     Refunded {
         htlc_location: L::HtlcLocation,
         deploy_transaction: L::Transaction,
         fund_transaction: L::Transaction,
         refund_transaction: L::Transaction,
+        asset: A,
     },
     IncorrectlyFunded {
         htlc_location: L::HtlcLocation,
         deploy_transaction: L::Transaction,
         fund_transaction: L::Transaction,
+        asset: A,
     },
 }
 
+impl<L: Ledger, A: Asset> LedgerState<L, A> {
+    pub fn transition_to_deployed(&mut self, deployed: Deployed<L>) {
+        let Deployed {
+            transaction,
+            location,
+        } = deployed;
+
+        match std::mem::replace(self, LedgerState::NotDeployed) {
+            LedgerState::NotDeployed => {
+                *self = LedgerState::Deployed {
+                    deploy_transaction: transaction,
+                    htlc_location: location,
+                }
+            }
+            other => panic!("expected state NotDeployed, got {}", HtlcState::from(other)),
```

Comment:
:thinking: I wish we could leverage the type system to avoid these panics. Hmmm.

Reply:
Don't really know if we can, these panics are bugs IMO so it should be fine.